internal/app/machined/pkg/runtime/v1alpha1/platform/metal/metal_test.go

```diff
@@ -48,19 +94,252 @@ func TestPopulateURLParameters(t *testing.T) {
 			url:         "http://example.com/metadata?uuid=xyz&uuid=foo",
 			expectedURL: fmt.Sprintf("http://example.com/metadata?uuid=%s", mockUUID),
 		},
+		{
+			name:        "single serial number",
+			url:         "http://example.com/metadata?serial=${serial}",
+			expectedURL: fmt.Sprintf("http://example.com/metadata?serial=%s", mockSerialNumber),
+		},
+		{
+			name:        "single MAC",
+			url:         "http://example.com/metadata?mac=${mac}",
+			expectedURL: fmt.Sprintf("http://example.com/metadata?mac=%s", mockMAC),
+		},
+		{
+			name:        "single hostname",
+			url:         "http://example.com/metadata?host=${hostname}",
+			expectedURL: fmt.Sprintf("http://example.com/metadata?host=%s", mockHostname),
+		},
+		{
+			name:        "serial number, MAC and hostname",
+			url:         "http://example.com/metadata?h=${hostname}&m=${mac}&s=${serial}",
+			expectedURL: fmt.Sprintf("http://example.com/metadata?h=%s&m=%s&s=%s", mockHostname, mockMAC, mockSerialNumber),
+		},
+		{
+			name:        "MAC and UUID without variable",
+			url:         "http://example.com/metadata?macaddr=${mac}&uuid=",
+			expectedURL: fmt.Sprintf("http://example.com/metadata?macaddr=%s&uuid=%s", mockMAC, mockUUID),
+		},
+		{
+			name:        "serial number and UUID without variable, order is not preserved",
+			url:         "http://example.com/metadata?uuid=&ser=${serial}",
+			expectedURL: fmt.Sprintf("http://example.com/metadata?ser=%s&uuid=%s", mockSerialNumber, mockUUID),
+		},
+		{
+			name:        "UUID variable",
+			url:         "http://example.com/metadata?uuid=${uuid}",
+			expectedURL: fmt.Sprintf("http://example.com/metadata?uuid=%s", mockUUID),
+		},
+		{
+			name:        "serial number and UUID with variable, order is not preserved",
+			url:         "http://example.com/metadata?uuid=${uuid}&ser=${serial}",
+			expectedURL: fmt.Sprintf("http://example.com/metadata?ser=%s&uuid=%s", mockSerialNumber, mockUUID),
+		},
 	} {
 		tt := tt
 
 		t.Run(tt.name, func(t *testing.T) {
-			output, err := metal.PopulateURLParameters(tt.url, func() (string, error) {
-				return mockUUID, nil
-			})
+			getSystemUUIDAndSerialNumber := func() (string, string, error) {
+				return mockUUID, mockSerialNumber, nil
+			}
+
+			getMACAddress := func(context.Context, state.State) (string, error) {
+				return mockMAC, nil
+			}
+
+			getHostname := func(context.Context, state.State) (string, error) {
+				return mockHostname, nil
+			}
+
+			output, err := metal.PopulateURLParameters(context.Background(), tt.url, nil, getSystemUUIDAndSerialNumber, getMACAddress, getHostname)
 
 			if tt.expectedError != "" {
 				assert.EqualError(t, err, tt.expectedError)
 			} else {
-				assert.Equal(t, output, tt.expectedURL)
+				assert.Equal(t, tt.expectedURL, output)
 			}
 		})
 	}
 }
+
+type mockState struct {
```

Comment:
I used a mock because I wanted to test the behaviour in [getResource](https://github.com/siderolabs/talos/blob/ffe94876ed7f6c629ea8e9ec5441bbbade972712/internal/app/machined/pkg/runtime/v1alpha1/platform/metal/metal.go#L169-L201), where it first calls state.State.List which returns nothing and later calls state.State.WatchKind. If I populate a real state.State I couldn't simulate this. Or maybe I didn't find a way. Is there an example where we do something similar?